smartpower3/screens/voltagescreen.cpp

```diff
@@ -0,0 +1,374 @@
+		if (mode == VOLTAGE_BASE_MOVE) {
+			channel[0]->setVolt(dial_cnt);  // this by default sets incremental difference to currently set value
+			if (dial_cnt != 0) {
+				settings->setChannel0Voltage(channel[0]->getVolt());
+			}
+		} else {
+			channel[0]->editVolt(dial_cnt);
+		}
+	} else if (selected == STATE_CURRENT0) {
+		clampVariableToRange(-(current_limit - 5), (30 - current_limit), &dial_cnt);
+		if (mode == VOLTAGE_BASE_MOVE) {
+			channel[0]->setCurrentLimit(dial_cnt);  // this by default sets incremental difference to currently set value
+			if (dial_cnt != 0) {
+				settings->setChannel0CurrentLimit(channel[0]->getCurrentLimit());
+			}
+		} else {
+			channel[0]->editCurrentLimit(dial_cnt);
+		}
+
+	} else if (selected == STATE_VOLT1) {
+		clampVoltageDialCountToRange(volt_set, &dial_cnt);
+		if (mode == VOLTAGE_BASE_MOVE) {
+			channel[1]->setVolt(dial_cnt);  // this by default sets incremental difference to currently set value
+			if (dial_cnt != 0) {
+				settings->setChannel1Voltage(channel[1]->getVolt());
+			}
+		} else {
+			channel[1]->editVolt(dial_cnt);
+		}
+	} else if (selected == STATE_CURRENT1) {
+		clampVariableToRange(-(current_limit - 5), (30 - current_limit), &dial_cnt);
+		if (mode == VOLTAGE_BASE_MOVE) {
+			channel[1]->setCurrentLimit(dial_cnt);  // this by default sets incremental difference to currently set value
+			if (dial_cnt != 0) {
+				settings->setChannel1CurrentLimit(channel[1]->getCurrentLimit());
+			}
+		} else {
+			channel[1]->editCurrentLimit(dial_cnt);
+		}
+	}
+}
+
+void VoltageScreen::checkSTPD01InterruptStat()
+{
+	for (int i = 0; i < 2; i++) {
+		channel[i]->isAvailableSTPD01();
+		int_stat[i] = channel[i]->checkInterruptStat(onoff[i]);
+	}
+}
+
+uint8_t VoltageScreen::getIntStat(uint8_t channel)
+{
+	return int_stat[channel];
+}
+
+void VoltageScreen::countLowVoltage(uint8_t ch)
+{
+	channel[ch]->countLowVoltage();
+}
+
+void VoltageScreen::clearLowVoltage(uint8_t ch)
+{
+	channel[ch]->clearLowVoltage();
+}
+
+Channel* VoltageScreen::getChannel(uint8_t channel_number)
+{
+	return this->channel[channel_number];
+}
+
+bool VoltageScreen::getEnabledSTPD01(uint8_t which_stpd)
+{
+	return this->enabled_stpd01[which_stpd];
+}
+
+void VoltageScreen::setEnabledSTPD01(uint8_t which_stpd, bool enabled)
+{
+	this->enabled_stpd01[which_stpd] = enabled;
+}
+
+void VoltageScreen::onShutdown(void)
+{
+	this->shutdown = true;
+	onoff[0] = 0;
+	onoff[1] = 0;
+	this->getChannel(0)->off();
+	this->getChannel(1)->off();
+	this->is_initialized = false;
+}
+
+void VoltageScreen::onWakeup(void)
+{
+	this->shutdown = false;
+	selected = dial_cnt = dial_cnt_old = STATE_NONE;
+}
+
+void VoltageScreen::onLeave(void)
+{
+	this->is_initialized = false;
+	channel[0]->setHide();
+	channel[1]->setHide();
+}
+
+void VoltageScreen::onEnter(void)
+{
+	channel[0]->clearHide();
+	channel[1]->clearHide();
+}
+
+screen_t VoltageScreen::getType(void)
+{
+	return VOLTAGE_SCREEN;
+}
```

Comment:
All screens should implement this, so it is possible to figure out from the code which screen we are on, because cpp does not have introspection features.

Inheritance - all screens are children of `Screen`, so it is possible to do something like [pseudo]:
``` 
Screen active_screen;
for (int i = LAST_SCREEN_COUNT; i > 0; i++) {
    active_screen = screens[i];
    active_screen->onShutdown();  // Do something to all the screens
    if (active_screen.getType() == VOLTAGE_SCREEN {
        ...do something else...
    }
}